# Model selection

Our task is to train a model that will be able to the binary classification whether a train is late or not. There are several models to do this supervised learning task with and in this notebook we will lay out advantages and disadvantages of different models.

Before evaluating how good different models fit a task it is important to see how much data we have and how it looks. A more thourough analysis of our dataset can be found in the [data exploration file](data_exploration.ipynb). 

In [ ]:
import pandas as pd
import numpy as np

data = pd.read_csv('../data/DBtrainrides.csv', encoding='UTF-8')

In [3]:
data.shape

(2061357, 20)

In [6]:
data.sort_values('departure_plan').head()

ID  line  \
1442  -8181182839806210915-2407072216-17    10   
3658  -9084944889004202938-2407072259-15  RB41   
1409   -834486270486097433-2407072244-31     6   
298    -1961474624078022131-2407072316-4    50   
3048  -1291355389794506596-2407072318-13    12   

                                                   path   eva_nr  category  \
1442  Doberlug-Kirchhain|Finsterwalde(Niederlausitz)...  8011471         5   
3658  Friedberg(Hess)|Bad Nauheim|Ostheim(b Butzbach...  8003856         5   
1409  Tutzing|Feldafing|Possenhofen|Starnberg|Starnb...  8001682         4   
298              Nürnberg Hbf|Neumarkt(Oberpf)|Parsberg  8000882         5   
3048  Blankenberg(Sieg)|Hennef im Siegbogen|Hennef(S...  8000207         1   

               station                state              city    zip  \
1442  Eisenhüttenstadt          Brandenburg  Eisenhüttenstadt  15890   
3658       Marburg Süd               Hessen           Marburg  35039   
1409        Eglharting               Bayern        Kirchseeon  85614   
298       Beratzhausen               Bayern      Beratzhausen  93176   
3048          Köln Hbf  Nordrhein-Westfalen              Köln  50667   

           long        lat         arrival_plan       departure_plan  \
1442  14.658435  52.147980  2024-07-07 23:59:00  2024-07-08 00:00:00   
3658   8.762431  50.794467  2024-07-07 23:59:00  2024-07-08 00:00:00   
1409  11.864291  48.082348  2024-07-07 23:59:00  2024-07-08 00:00:00   
298   11.807853  49.092846  2024-07-07 23:59:00  2024-07-08 00:00:00   
3048   6.958729  50.943030  2024-07-07 23:59:00  2024-07-08 00:00:00   

           arrival_change     departure_change  arrival_delay_m  \
1442  2024-07-07 23:59:00  2024-07-08 00:00:00                0   
3658  2024-07-07 23:59:00  2024-07-08 00:00:00                0   
1409  2024-07-08 00:00:00  2024-07-08 00:00:00                1   
298   2024-07-08 00:16:00  2024-07-08 00:17:00               17   
3048  2024-07-08 00:01:00  2024-07-08 00:02:00                2   

      departure_delay_m         info arrival_delay_check departure_delay_check  
1442                  0          NaN             on_time               on_time  
3658                  0          NaN             on_time               on_time  
1409                  0  Bauarbeiten             on_time               on_time  
298                  17          NaN               delay                 delay  
3048                  2          NaN             on_time               on_time

In [9]:
data.sort_values('departure_plan').tail()

ID  line  \
2055119   -7723223610149163054-2407142310-7    69   
2055959   6187565811835269028-2407142314-10  RB69   
2052851  -7113663931471534874-2407142315-24     4   
2055117   -9088865381407852178-2407142310-7    89   
2052903   -2117726194312836395-2407142308-7    71   

                                                      path   eva_nr  category  \
2055119  Münster(Westf)Hbf|Münster-Hiltrup|Rinkerode|Dr...  8000149         2   
2055959  Augsburg Hbf|Augsburg Morellstr.|Augsburg Mess...  8000195         4   
2052851  Karlsruhe Albtalbahnhof|Karlsruhe Bahnhofsvorp...  8000053         4   
2055117  Münster(Westf)Hbf|Münster-Hiltrup|Rinkerode|Dr...  8000149         2   
2052903  Memmingen|Sontheim(Schwab)|Stetten(Schwab)|Min...  8000057         3   

                  station                state       city    zip       long  \
2055119  Hamm (Westf) Hbf  Nordrhein-Westfalen      Hamm   59065   7.807821   
2055959         Kaufering               Bayern  Kaufering  86916  10.854595   
2052851           Bretten    Baden-Württemberg    Bretten  75015   8.693448   
2055117  Hamm (Westf) Hbf  Nordrhein-Westfalen      Hamm   59065   7.807821   
2052903           Buchloe               Bayern    Buchloe  86807  10.716217   

               lat         arrival_plan       departure_plan  \
2055119  51.678078  2024-07-14 23:37:00  2024-07-15 00:07:00   
2055959  48.080835  2024-07-14 23:52:00  2024-07-15 00:07:00   
2052851  49.036900  2024-07-14 23:58:00  2024-07-15 00:08:00   
2055117  51.678078  2024-07-14 23:37:00  2024-07-15 00:09:00   
2052903  48.033723  2024-07-14 23:50:00  2024-07-15 00:10:00   

              arrival_change     departure_change  arrival_delay_m  \
2055119                  NaN                  NaN                0   
2055959  2024-07-14 23:52:00  2024-07-15 00:07:00                0   
2052851  2024-07-14 23:58:00  2024-07-15 00:08:00                0   
2055117  2024-07-14 23:37:00  2024-07-15 00:09:00                0   
2052903  2024-07-14 23:51:00  2024-07-15 00:11:00                1   

         departure_delay_m info arrival_delay_check departure_delay_check  
2055119                  0  NaN             on_time               on_time  
2055959                  0  NaN             on_time               on_time  
2052851                  0  NaN             on_time               on_time  
2055117                  0  NaN             on_time               on_time  
2052903                  1  NaN             on_time               on_time

In [4]:
data.describe()

eva_nr      category           zip          long           lat  \
count  2.061357e+06  2.061357e+06  2.061357e+06  2.061357e+06  2.061357e+06   
mean   8.018267e+06  3.871748e+00  4.627924e+04  1.018367e+01  5.088240e+01   
std    3.178659e+04  1.037063e+00  2.821424e+04  2.273525e+00  1.792217e+00   
min    8.000001e+06  1.000000e+00  1.067000e+03  6.070715e+00  4.741103e+01   
25%    8.001582e+06  3.000000e+00  1.810900e+04  8.494709e+00  4.935329e+01   
50%    8.004136e+06  4.000000e+00  4.705100e+04  9.944088e+00  5.108746e+01   
75%    8.010208e+06  5.000000e+00  7.080600e+04  1.209055e+01  5.247854e+01   
max    8.098360e+06  5.000000e+00  9.997400e+04  1.497908e+01  5.490684e+01   

       arrival_delay_m  departure_delay_m  
count     2.061357e+06       2.061357e+06  
mean      1.176581e+00       1.223574e+00  
std       3.407859e+00       3.418300e+00  
min       0.000000e+00       0.000000e+00  
25%       0.000000e+00       0.000000e+00  
50%       0.000000e+00       0.000000e+00  
75%       1.000000e+00       1.000000e+00  
max       1.590000e+02       1.590000e+02

In [6]:
for col in data.columns:
    if not pd.api.types.is_numeric_dtype(data[col]):
        print(f'{col}: {len(data[col].unique())}')
        #print(f'{col}: {data[col].unique()}')

ID: 2029894
line: 296
path: 22154
station: 1996
state: 16
city: 1292
arrival_plan: 10085
departure_plan: 10089
arrival_change: 10123
departure_change: 10119
info: 8
arrival_delay_check: 2
departure_delay_check: 2


In [15]:
print(data[data['arrival_delay_check'] == 'on_time'].shape[0], data[data['arrival_delay_check'] == 'on_time'].shape[0]/data.shape[0])
print(data[data['arrival_delay_check'] == 'delay'].shape[0], data[data['arrival_delay_check'] == 'delay'].shape[0]/data.shape[0])

1950404 0.9461747771007157
110953 0.053825222899284306


One thing we should keep in mind is that our dataset is heavily imbalanced with only ~5% of our data set being delayed. We therefore should use something like the f-score or just the recall as our measurement for how well a given model has performed. 

Due to the amount of data it is also hard for us too say upfront which parameters are going to perform well and which won't. Since we have a lot of data points available we will opt to use k-fold cross validation to find the best parameter combination for the model we choose.

In order to deal with the imbalance in our dataset we will just sample the same amount of trains that were on time, as we have trains that were delayed and directly put the other trains that were on time into the test set.

## Decision Tree Classifier

Decision trees have the advantage that they are easy to understand and can be interpreted easily as well. However they are often a bit simple and suffer more easily from overfitting than other models.

## Random Forest Classifier

Random Forest Classifiers are probably a better choice than one single decision tree, because they are less prone too overfitting if enough data points are available, which is given in our case with 2061357 datapoints.

## Gradient Boosting

Gradient Boosting is most likely a better fit than a random forest classifier. The advantages of this model are that it will probably be able to achieve a high accuracy. The problem is that we will most likely need to do a lot of cross validation for the hyperparameters and this will take a lot of time due to the nature of the model.

## K-Nearest-Neighbor 

KNN is also a possible choice. Since we have significantly more data points than features (~2*10⁷ >> 20), it is unlikely that the course of dimensionality is going to be a big problem in our case. We will most likely use only ~10% of the datapoints available for training but these are still way more datapoints. One problem we might encounter is that we should probably switch to one-hot-encoding for different nominal values too be able too use them. This will blow up the amount of features that we have but it will most likely still be less than the amount of data points we have. But this is definetly something to keep in mind. 

If we see that we don't have enough datapoints we can still use PCA to reduce dimensionality again.

## Neural Network

A neural network is possibly a very accurate solution as we have a lot of datapoints and can therefore train the model very long. However we do think it is not necessary tooo train a neural network as other more efficient options are also available that will probably reach almost the same accuracy as a neural network.

## Support Vector Machine

We don't think that SVMs are a suitable approach to our problem because if we use one hot encoding we have a lot of dimensions and a lot of data points which will make the use of a SVM very complex.

## Conclusion

We will run some smaller tests on the efficiency and accuracy/f-score of a gradient boosting model. In case of promising results we will continue too use them. Otherwise we will choose to go for a KNN model since we deem it to be the next best fit because we have a lot of data available and it is way less time consuming allowing us to try out more hyperparameters. In case this model does not provide a satisfying f-score either, we will try to apply PCA or go for a neural network instead.